transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,317 @@
+                            }
+                            ++acquiredChannelCount;
+                            --pendingAcquireCount;
+
+                            FixedChannelPool.super.acquire(task.key, task.promise);
+                        }
+                    }
+                };
+                break;
+            default:
+                throw new Error();
+            }
+        }
+        executor = bootstrap.group().next();
+        this.maxConnections = maxConnections;
+        this.maxPendingAcquires = maxPendingAcquires;
+    }
+
+    @Override
+    public Future<C> acquire(final K key, final Promise<C> promise) {
+        try {
+            final Promise<C> p = executor.newPromise();
+            if (executor.inEventLoop()) {
+                acquire0(key, promise, p);
+            } else {
+                executor.execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        acquire0(key, promise, p);
+                    }
+                });
+            }
+            return p;
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void acquire0(K key, final Promise<C> originalPromise, Promise<C> p) {
+        assert executor.inEventLoop();
+
+        p.addListener(new FutureListener<C>() {
+            @Override
+            public void operationComplete(Future<C> future) throws Exception {
+                assert executor.inEventLoop();
+
+                if (future.isSuccess()) {
+                    originalPromise.setSuccess(future.getNow());
+                } else {
+                    // Something went wrong try to run pending acquire tasks.
+                    --acquiredChannelCount;
+
+                    assert acquiredChannelCount > 0;
+
+                    // Run the pending acquire tasks before notify the original promise so if the user would
+                    // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+                    // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+                    // more.
+                    runTaskQueue();
+                    originalPromise.setFailure(future.cause());
+                }
+            }
+        });
+        if (acquiredChannelCount < maxConnections) {
+            ++acquiredChannelCount;
+
+            assert acquiredChannelCount > 0;
+
+            super.acquire(key, p);
+        } else {
+            if (++pendingAcquireCount > maxPendingAcquires) {
+                originalPromise.setFailure(FULL_EXCEPTION);
+            } else {
+                pendingAcquireQueue.offer(new AcquireTask<C, K>(key, p));
+                if (timeoutTask != null) {
+                    executor.schedule(timeoutTask, acquireTimeoutNanos, TimeUnit.NANOSECONDS);
+                }
+            }
+
+            assert pendingAcquireCount > 0;
```

Comment:
Can you add a comment as to why this is the case?  Seems strange to increment the count event if we are not offering it to the queue.